test/gtest/common/test_helpers.h

```diff
@@ -476,6 +480,80 @@ class message_stream {
 } // ucs
 
 
+namespace ucp {
+
+class data_type_desc_t {
+public: 
+    data_type_desc_t() :  _origin(0), _length(0), _buf(NULL) {};
+
+    data_type_desc_t(ucp_datatype_t datatype, void *buf, size_t length)
+        : _origin(uintptr_t(buf)), _length(length), _buf(NULL) {
+        make(datatype, buf, length);
+    }
+
+    data_type_desc_t(ucp_datatype_t datatype, void *buf, size_t length,
+                     size_t iov_count)
+        : _origin(uintptr_t(buf)), _length(length), _buf(NULL)
+    {
+        make(datatype, buf, length, iov_count);
+    };
+
+    data_type_desc_t &make(ucp_datatype_t datatype, void *buf, size_t length,
+                           size_t iov_count);
+
+    data_type_desc_t &make(ucp_datatype_t datatype, void *buf, size_t length) {
+        return make(datatype, buf, length, _iov_cnt_limit);
+    };
+
+    data_type_desc_t &forward_to(size_t offset) {
+        EXPECT_LE(offset, _length);
+        invalidate();
+        return make(_dt, (void *)(_origin + offset), _length - offset,
+                    _iov_cnt_limit);
+    };
+
+    ucp_datatype_t dt() const {
+        EXPECT_TRUE(is_valid());
+        return _dt;
+    };
+
+    void *buf() const {
+        EXPECT_TRUE(is_valid());
+        return _buf;
+    };
+
+    size_t count() const {
+        EXPECT_TRUE(is_valid());
+        return _count;
+    };
+
+    bool is_valid() const {
+        return (_buf != NULL) && (_count != 0) &&
+               (UCP_DT_IS_IOV(_dt) ? (_count <= _iov_cnt_limit) :
+                UCP_DT_IS_CONTIG(_dt));
+    }
+
+private:
+    void invalidate() {
+        EXPECT_TRUE(is_valid());
+        _buf   = NULL;
+        _count = 0;
+    }
+
+    uintptr_t      _origin;
+    size_t         _length;
+
+    ucp_datatype_t _dt;
+    void           *_buf;
+    size_t         _count;
+
+    static const size_t _iov_cnt_limit = 40;
```

Comment:
it's a tricky construction: use const variable to initialize array size, compilation may fail by paranoic compiler